src/wp-includes/fonts/class-wp-font-collection.php

```diff
@@ -0,0 +1,260 @@
+<?php
+/**
+ * Font Collection class.
+ *
+ * This file contains the Font Collection class definition.
+ *
+ * @package    WordPress
+ * @subpackage Fonts
+ * @since      6.5.0
+ */
+
+/**
+ * Font Collection class.
+ *
+ * @since 6.5.0
+ *
+ * @see wp_register_font_collection()
+ */
+final class WP_Font_Collection {
+	/**
+	 * The unique slug for the font collection.
+	 *
+	 * @since 6.5.0
+	 * @var string
+	 */
+	public $slug;
+
+	/**
+	 * Font collection data.
+	 *
+	 * @since 6.5.0
+	 * @var array|WP_Error|null
+	 */
+	private $data;
+
+	/**
+	 * Font collection JSON file path or URL.
+	 *
+	 * @since 6.5.0
+	 * @var string|null
+	 */
+	private $src;
+
+	/**
+	 * WP_Font_Collection constructor.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string        $slug         Font collection slug.
+	 * @param array|string  $data_or_file Font collection data array or a path/URL to a JSON file
+	 *                                    containing the font collection.
+	 *                                    See {@see wp_register_font_collection()} for the supported fields.
+	 */
+	public function __construct( $slug, $data_or_file ) {
+		$this->slug = sanitize_title( $slug );
+
+		if ( is_array( $data_or_file ) ) {
+			$this->data = $this->sanitize_and_validate_data( $data_or_file );
+		} else {
+			// JSON data is lazy loaded by ::get_data().
+			$this->src = $data_or_file;
+		}
+
+		if ( $this->slug !== $slug ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: Font collection slug. */
+				sprintf( __( 'Font collection slug "%s" is not valid. Slugs must use only alphanumeric characters, dashes, and underscores.' ), $slug ),
+				'6.5.0'
+			);
+		}
```

Comment:
>Shouldn't this block be moved up, right after the call to sanitize_title?

+1 It'll keep the slug code together.

Reply:
> 1. Shouldn't this block be moved up, right after the call to sanitize_title?
We have a [PR started in Gutenberg](https://github.com/WordPress/gutenberg/pull/58691) to address feedback, including moving this check up.

> Is _doing_it_wrong enough, or should it throw? ...
Great question, it deserves discussion!
In searching through Core, it seems that throwing an exception is somewhat rare outside of included library code. Also, to me throwing seems extreme here. This code path is potentially loaded on every request when a font collection is registered: and erroring the whole site doesn't seem appropriate for a slug with extraneous characters. But if there's guidance for a more Core-like way to handle this, I'd appreciate it!

> ... Or assign a wp_error to $this->data or something?
`$slug` is not part of the data, so I don't think an error there makes sense.

Reply:
A few questions before assessing if throwing an exception vs doing it wrong.

How critical is this check?
```
if ( $this->slug !== $slug )
```

Is it critical that the sanitized slug exactly match the given slug?

What happens if the 2 don't match? Does the collection not work?

If it's critical to functionality, then throwing an exception _might_ be the way to go. If it's not critical and everything will continue to work properly, then it's informational.

If it's information, I then wonder: Is a doing it wrong necessary?

Reply:
hey @creativecoder, want to open an issue for this in Gutenberg? Then the discussion can move to there. Thinking this one needs more discussion and consideration, but does not need to be a blocker for this PR to land in Core.

Reply:
> What happens if the 2 don't match? Does the collection not work?

The collection will be registered and work, but a few potentially unexpected things happen. For example, if I use the slug `invâlíd / slüg`, e.g. `wp_register_font_collection( 'invâlíd / slüg', ... )`

- The `$collection->slug` value is now the sanitized value `invalid-slug` rather than `invâlíd / slüg`.
- `WP_Font_Library::get_font_collection( 'invâlíd / slüg' )` returns an error, because a collection with that slug is not registered.
- `wp/v2/font-collections/invâlíd / slüg` (or it's uri encoded version) will return an error, because a collection with that slug is not found.

However, `WP_Font_Library::get_font_collections` and `wp/v2/font-collections` still returns the collection, but with a slug that is sanitized (`invalid-slug`). Additionally
- `WP_Font_Library::get_font_collection( 'invalid-slug' )` returns the collection without error.
- `wp/v2/font-collections/invalid-slug` returns the collection without error.

Reply:
Since it still works. thinking throwing an exception is too severe. The `_doing_it_wrong()` informs just in case unexpected things happen. Currently thinking it's fine as is.

That said, I still suggest open a GB issue to capture this discussion and allow it to continue. That way its consideration is captured and any follow-up can happen if changes are needed (thus unblocking this PR).

Reply:
> hey @creativecoder, want to open an issue for this in Gutenberg?

Good idea! Issue opened here: https://github.com/WordPress/gutenberg/issues/58695

Reply:
Folks, moving this discussion over to https://github.com/WordPress/gutenberg/issues/58695. It's a great discussion to have. Thanks @mcsf for starting it.

As this discussion has moved, it should not block this PR. If there are changes from the discussion, those change can be handled in a follow-up Core patch and commit.